core/src/bitmap/operations.rs

```diff
@@ -1201,6 +1201,168 @@ pub fn apply_filter<'gc>(
     }
 }
 
+#[allow(clippy::too_many_arguments)]
+fn copy_on_cpu<'gc>(
+    context: &mut UpdateContext<'_, 'gc>,
+    source: BitmapDataWrapper<'gc>,
+    dest: BitmapDataWrapper<'gc>,
+    source_region: PixelRegion,
+    dest_region: PixelRegion,
+    mut blend: bool,
+) {
+    if !source.transparency() {
+        // We don't need to blend if we're copying an opaque texture, the alpha would be 255 anyway
+        blend = false;
+    }
+    if !blend && source.ptr_eq(dest) && source_region == dest_region {
+        // Copying the same area of self to self, noop
+        return;
+    }
+
+    if source.ptr_eq(dest) {
+        let dest = dest.sync();
+        let mut write = dest.write(context.gc_context);
+
+        for y in 0..dest_region.height() {
+            for x in 0..dest_region.width() {
+                let mut color =
+                    write.get_pixel32_raw(source_region.x_min + x, source_region.y_min + y);
+                if blend {
+                    color = write
+                        .get_pixel32_raw(dest_region.x_min + x, dest_region.y_min + y)
+                        .blend_over(&color);
+                }
+                write.set_pixel32_raw(dest_region.x_min + x, dest_region.y_min + y, color);
+            }
+        }
+
+        write.set_cpu_dirty(dest_region);
+    } else {
+        let dest = dest.sync();
+        let mut dest_write = dest.write(context.gc_context);
+        let source_read = source.read_area(source_region);
+
+        if !blend && (dest_write.transparency() || !source_read.transparency()) {
+            // Copying anything to a transparent texture,
+            // or copying an opaque texture to an opaque texture,
+            // means we can skip alpha premultiplication
+
+            if dest_region == source_region
+                && dest_region.width() == dest_write.width()
+                && dest_region.height() == dest_write.height()
+                && dest_region.width() == source_read.width()
+                && dest_region.height() == source_read.height()
+            {
+                // Copying an entire texture that's the same size and type? Just replace the whole thing
+                *dest_write.raw_pixels_mut() = source_read.raw_pixels().to_owned();
+            } else {
+                for y in 0..dest_region.height() {
+                    for x in 0..dest_region.width() {
+                        let color = source_read
+                            .get_pixel32_raw(source_region.x_min + x, source_region.y_min + y);
+                        dest_write.set_pixel32_raw(
+                            dest_region.x_min + x,
+                            dest_region.y_min + y,
+                            color,
+                        );
+                    }
+                }
+            }
+        } else {
+            // Copying a transparent texture to an opaque texture, or blending
```

Comment:
Same comment nit as above, this code path also happens when copying transparent onto transparent with `mergeAlpha`.